pkg/k8s/informers/informers.go

```diff
@@ -0,0 +1,191 @@
+)
+
+type InformerCollectionOption func(*InformerCollection)
+
+// NewInformerCollection creates a new InformerCollection
+func NewInformerCollection(meshName string, stop <-chan struct{}, opts ...InformerCollectionOption) (*InformerCollection, error) {
+
+	ic := &InformerCollection{
+		meshName:  meshName,
+		informers: map[InformerKey]*informer{},
+	}
+
+	for _, opt := range opts {
+		opt(ic)
+	}
+
+	// Initialize informers
+	informerInitHandlerMap := map[InformerKey]informerInit{
+		// Kubernetes
+		InformerKeyNamespace:      ic.initNamespaceMonitor,
+		InformerKeyService:        ic.initServicesMonitor,
+		InformerKeyServiceAccount: ic.initServiceAccountsMonitor,
+		InformerKeyPod:            ic.initPodMonitor,
+		InformerKeyEndpoints:      ic.initEndpointMonitor,
+
+		// SMI
+		InformerKeyTrafficSplit:   ic.initTrafficSplitMonitor,
+		InformerKeyTrafficTarget:  ic.initTrafficTargetMonitor,
+		InformerKeyHTTPRouteGroup: ic.initHTTPRouteGroupMonitor,
+		InformerKeyTCPRoute:       ic.initTCPRouteMonitor,
+
+		// Config
+		InformerKeyMeshConfig:          ic.initMeshConfigMonitor,
+		InformerKeyMeshRootCertificate: ic.initMeshRootCertificateMonitor,
+
+		// Policy
+		InformerKeyEgress:                 ic.initEgressMonitor,
+		InformerKeyIngressBackend:         ic.initIngressBackendMonitor,
+		InformerKeyUpstreamTrafficSetting: ic.initUpstreamTrafficSettingMonitor,
+		InformerKeyRetry:                  ic.initRetryMonitor,
+	}
+
+	if len(ic.selectedInformers) == 0 {
+		// Initialize all informers
+		ic.selectedInformers = []InformerKey{
+			InformerKeyNamespace,
+			InformerKeyService,
+			InformerKeyPod,
+			InformerKeyEndpoints,
+			InformerKeyServiceAccount,
+			InformerKeyTrafficSplit,
+			InformerKeyTrafficTarget,
+			InformerKeyHTTPRouteGroup,
+			InformerKeyTCPRoute,
+			InformerKeyMeshConfig,
+			InformerKeyMeshRootCertificate,
+			InformerKeyEgress,
+			InformerKeyIngressBackend,
+			InformerKeyUpstreamTrafficSetting,
+			InformerKeyRetry,
+		}
+	}
+
+	for _, key := range ic.selectedInformers {
+		informerInitHandlerMap[key]()
+	}
+
+	if err := ic.run(stop); err != nil {
+		log.Error().Err(err).Msg("Could not start informer collection")
+		return nil, err
+	}
+
+	return ic, nil
+}
+
+// WithCustomStore sets the shared store for the InformerCollection to reference.
+// This store will be used instead of each informer's store. This should typically
+// only be used in tests
+func WithCustomStores(stores map[InformerKey]cache.Store) InformerCollectionOption {
+	return func(ic *InformerCollection) {
+		ic.customStores = stores
+	}
+}
+
+// WithSelectedInformers sets the selected informers for the InformerCollection
+func WithSelectedInformers(selectedInformers ...InformerKey) InformerCollectionOption {
+	return func(ic *InformerCollection) {
+		ic.selectedInformers = selectedInformers
+	}
+}
+
+// WithKubeClient sets the kubeClient for the InformerCollection
+func WithKubeClient(kubeClient kubernetes.Interface) InformerCollectionOption {
```

Comment:
Is there ever a place where we pass in a client but don't run the informer? If not, I'd say this could be the signal.

If there is, maybe we could add a bool here, runInformer? Simplifying the API otherwise you could specify an informer to run where the client was never passed in, which would be a bug.

Reply:
a further suggestion: put the init methods in here and get rid of the initInformerMap

Reply:
I think the map makes initialization cleaner; the alternatives are to re-declare a subset of the map and loop through the keys again or to have each method called sequentially. Happy to discuss further, but I feel like it's probably a preference